plugin/plugins/filesystem/nodes/root.go

```diff
@@ -0,0 +1,203 @@
+package fsnodes
+
+import (
+	"context"
+
+	"github.com/djdv/p9/p9"
+	cid "github.com/ipfs/go-cid"
+	nodeopts "github.com/ipfs/go-ipfs/plugin/plugins/filesystem/nodes/options"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	"github.com/multiformats/go-multihash"
+)
+
+var _ p9.File = (*RootIndex)(nil)
+var _ walkRef = (*RootIndex)(nil)
+
+const nRoot = "" // root namespace is intentionally left blank
+
+type rootPath string
+
+func (rp rootPath) String() string { return string(rp) }
+func (rootPath) Namespace() string { return nRoot }
+func (rootPath) Mutable() bool     { return true }
+func (rootPath) IsValid() error    { return nil }
+func (rootPath) Root() cid.Cid     { return cid.Cid{} }
+func (rootPath) Remainder() string { return "" }
+func (rp rootPath) Cid() cid.Cid {
+	prefix := cid.V1Builder{Codec: cid.DagCBOR, MhType: multihash.BLAKE2B_MIN}
+	c, err := prefix.Sum([]byte(rp))
+	if err != nil {
+		panic(err) //invalid root
+	}
+	return c
+}
+
+type systemTuple struct {
+	file   walkRef
+	dirent p9.Dirent
+}
+
+//TODO: rename, while this is likely to be the root, it doesn't have to be; maybe "IPFSOverlay"
+// RootIndex is a virtual directory file system, that maps a set of file system implementations to a hierarchy
+// Currently: "/ipfs":PinFS, "/ipfs/*:IPFS
+type RootIndex struct {
+	IPFSBase
+	subsystems map[string]systemTuple
+}
+
+// OverlayFileMeta holds data relevant to file system nodes themselves
+type OverlayFileMeta struct {
```

Comment:
I'll need to make a pass on all of the components that make up a filesystem object.
I'm not yet sure if some of these components would be useful at the package level.
The overlay metadata is probably worth considering as it would allow people to implement objects that can be utilized by our root index / the overlay. The rest of the metadata may be useful as well but needs some more consideration.

***
Thinking out loud:

Essentially a file system object is comprised of the `Base` class which holds 9P relevant file system metadata (such as a file's attributes, a node's path "trail" to track its location), the bare minimum to have a path based file system.
We add on to that to make an `IPFSBase` which should add the minimum requirements to interact with an IPFS node / the CoreAPI.
From here, every filesystem is then expected to add its own data, for IPFS this means things like file descriptors for open references, for MFS, the root, etc. and these are grouped into `xFileMetadata`

This is seperated somewhat nicely for us allowing you to make a new FS by simply inheriting `IPFSBase` and adding your own storage space
```go
type MyFs struct {
    IPFSBase
    MyMetadata
}
```

But this implies a lot of inheritance rather than it being explicit, separated, and more thought out in regards to the public interface.
I'll have to look at this in a more wholesome way soon. Opinions welcomed as always too.